File: mvp-java/ui/src/main/java/com/wasisto/githubuserfinder/android/ui/search/SearchActivity.java
package com.wasisto.githubuserfinder.android.ui.search;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.RecyclerView;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import com.wasisto.githubuserfinder.android.DefaultServiceLocator;
import com.wasisto.githubuserfinder.android.R;
import com.wasisto.githubuserfinder.android.ServiceLocator;
import com.wasisto.githubuserfinder.android.ui.userdetails.UserDetailsActivity;
import com.wasisto.githubuserfinder.domain.models.SearchHistoryItem;
import com.wasisto.githubuserfinder.domain.models.SearchUsersResult;
import com.wasisto.githubuserfinder.domain.usecases.GetSearchHistoryUseCase;
import com.wasisto.githubuserfinder.domain.usecases.SearchUsersUseCase;

import java.util.List;

public class SearchActivity extends AppCompatActivity implements SearchView {

    private SearchPresenter presenter;

    private ProgressBar loadingIndicator;

    private EditText queryEditText;

    private Button searchButton;

    private RecyclerView searchResultRecyclerView;

    private RecyclerView searchHistoryRecyclerView;

    private TextView noResultsTextView;

    private SearchResultAdapter searchResultAdapter;

    private SearchHistoryAdapter searchHistoryAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_search);

        ServiceLocator serviceLocator = DefaultServiceLocator.getInstance(getApplication());

        presenter = new SearchPresenter(
                this,
                new SearchUsersUseCase(
                        serviceLocator.getGithubDataSource(),
                        serviceLocator.getSearchHistoryDataSource(),
                        serviceLocator.getLogger(SearchUsersUseCase.class)
                ),
                new GetSearchHistoryUseCase(
                        serviceLocator.getSearchHistoryDataSource()
                ),
                serviceLocator.getLogger(SearchPresenter.class),
                serviceLocator.getIoExecutorService(),
                serviceLocator.getMainExecutorService()
        );

        loadingIndicator = findViewById(R.id.loadingIndicator);
        queryEditText = findViewById(R.id.queryEditText);
        searchButton = findViewById(R.id.searchButton);
        searchResultRecyclerView = findViewById(R.id.searchResultRecyclerView);
        searchHistoryRecyclerView = findViewById(R.id.searchHistoryRecyclerView);
        noResultsTextView = findViewById(R.id.noResultsTextView);

        searchResultAdapter = new SearchResultAdapter(presenter);
        searchResultRecyclerView.setAdapter(searchResultAdapter);

        searchHistoryAdapter = new SearchHistoryAdapter(presenter);
        searchHistoryRecyclerView.setAdapter(searchHistoryAdapter);

        searchButton.setOnClickListener(v -> presenter.search(queryEditText.getText().toString()));

        presenter.load();
    }

    @Override
    public void showLoadingIndicator() {
        loadingIndicator.setVisibility(View.VISIBLE);
    }

    @Override
    public void hideLoadingIndicator() {
        loadingIndicator.setVisibility(View.GONE);
    }

    @Override
    public void setQuery(String query) {
        queryEditText.setText(query);
    }

    @Override
    public void showSearchResult(List<SearchUsersResult.Item> searchResult) {
        searchResultRecyclerView.setVisibility(View.VISIBLE);

        searchResultAdapter.setData(searchResult);
        searchResultAdapter.notifyDataSetChanged();
    }

    @Override
    public void showSearchHistory(List<SearchHistoryItem> searchHistory) {
        searchHistoryRecyclerView.setVisibility(View.VISIBLE);

        searchHistoryAdapter.setData(searchHistory);
        searchHistoryAdapter.notifyDataSetChanged();
    }

    @Override
    public void hideSearchHistory() {
        searchHistoryRecyclerView.setVisibility(View.GONE);
    }

    @Override
    public void showNoResultsText() {
        noResultsTextView.setVisibility(View.VISIBLE);
    }

    @Override
    public void hideNoResultsText() {
        noResultsTextView.setVisibility(View.GONE);
    }

    @Override
    public void openUserDetailsActivity(String username) {
        startActivity(
                new Intent(this, UserDetailsActivity.class) {{
                    putExtra(UserDetailsActivity.EXTRA_USERNAME, username);
                }}
        );
    }

    @Override
    public void showErrorToast() {
        Toast.makeText(this, R.string.an_error_occurred, Toast.LENGTH_SHORT).show();
    }
}

File: mvp-kotlin/ui/src/main/java/com/wasisto/githubuserfinder/android/ui/search/SearchActivity.kt
package com.wasisto.githubuserfinder.android.ui.search

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import com.wasisto.githubuserfinder.android.DefaultServiceLocator
import com.wasisto.githubuserfinder.android.R
import com.wasisto.githubuserfinder.android.ui.userdetails.UserDetailsActivity
import com.wasisto.githubuserfinder.domain.models.SearchHistoryItem
import com.wasisto.githubuserfinder.domain.models.SearchUsersResult
import com.wasisto.githubuserfinder.domain.usecases.GetSearchHistoryUseCase
import com.wasisto.githubuserfinder.domain.usecases.SearchUsersUseCase
import kotlinx.android.synthetic.main.activity_search.*
import kotlinx.coroutines.Dispatchers

class SearchActivity : AppCompatActivity(), SearchView {

    private lateinit var presenter: SearchPresenter

    private lateinit var searchResultAdapter: SearchResultAdapter

    private lateinit var searchHistoryAdapter: SearchHistoryAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_search)

        val serviceLocator = DefaultServiceLocator.getInstance(application)

        presenter = SearchPresenter(
            this,
            SearchUsersUseCase(
                serviceLocator.githubDataSource,
                serviceLocator.searchHistoryDataSource,
                serviceLocator.getLogger(SearchUsersUseCase::class)
            ),
            GetSearchHistoryUseCase(
                serviceLocator.searchHistoryDataSource
            ),
            serviceLocator.getLogger(SearchPresenter::class),
            serviceLocator.ioDispatcher,
            serviceLocator.mainDispatcher,
            lifecycleScope
        )

        searchResultAdapter = SearchResultAdapter(presenter)
        searchResultRecyclerView.adapter = searchResultAdapter

        searchHistoryAdapter = SearchHistoryAdapter(presenter)
        searchHistoryRecyclerView.adapter = searchHistoryAdapter

        searchButton.setOnClickListener { presenter.search(queryEditText.text.toString()) }

        presenter.load()
    }

    override fun showLoadingIndicator() {
        loadingIndicator.visibility = View.VISIBLE
    }

    override fun hideLoadingIndicator() {
        loadingIndicator.visibility = View.GONE
    }

    override fun setQuery(query: String?) {
        queryEditText.setText(query)
    }

    override fun showSearchResult(searchResult: List<SearchUsersResult.Item>) {
        searchResultRecyclerView.visibility = View.VISIBLE
        searchResultAdapter.data = searchResult
        searchResultAdapter.notifyDataSetChanged()
    }

    override fun showSearchHistory(searchHistory: List<SearchHistoryItem>) {
        searchHistoryRecyclerView.visibility = View.VISIBLE
        searchHistoryAdapter.data = searchHistory
        searchHistoryAdapter.notifyDataSetChanged()
    }

    override fun hideSearchHistory() {
        searchHistoryRecyclerView.visibility = View.GONE
    }

    override fun showNoResultsText() {
        noResultsTextView.visibility = View.VISIBLE
    }

    override fun hideNoResultsText() {
        noResultsTextView.visibility = View.GONE
    }

    override fun openUserDetailsActivity(username: String) {
        startActivity(
            Intent(this, UserDetailsActivity::class.java).apply {
                putExtra(UserDetailsActivity.EXTRA_USERNAME, username)
            }
        )
    }

    override fun showErrorToast() {
        Toast.makeText(this, R.string.an_error_occurred, Toast.LENGTH_SHORT).show()
    }
}